usr.sbin/npf/npfctl/npf_altq.c

```diff
@@ -0,0 +1,1395 @@
+#include <unistd.h>
+#include "npf.h"
+#include <altq/altq.h>
+#include <altq/altq_cbq.h>
+#include <altq/altq_priq.h>
+#include <altq/altq_hfsc.h>
+#include "npfctl.h"
+
+LIST_HEAD(gen_sc, segment) rtsc, lssc;
+static int	eval_npfqueue_cbq(struct npf_altq *);
+static int	cbq_compute_idletime(struct npf_altq *);
+static int	check_commit_cbq(struct npf_altq *);
+static int	print_cbq_opts(const struct npf_altq *);
+static int	eval_npfqueue_priq(struct npf_altq *);
+static int	check_commit_priq(struct npf_altq *);
+static int	print_priq_opts(const struct npf_altq *);
+static int	eval_npfqueue_hfsc(struct npf_altq *);
+static int	check_commit_hfsc(struct npf_altq *);
+static int	print_hfsc_opts(const struct npf_altq *,
+		    const struct node_queue_opt *);
+static void		 gsc_add_sc(struct gen_sc *, struct service_curve *);
+static int		 is_gsc_under_sc(struct gen_sc *,
+			     struct service_curve *);
+static void		 gsc_destroy(struct gen_sc *);
+static struct segment	*gsc_getentry(struct gen_sc *, double);
+static int		 gsc_add_seg(struct gen_sc *, double, double, double,
+			     double);
+static double		 sc_x2y(struct service_curve *, double);
+void		 print_hfsc_sc(const char *, u_int, u_int, u_int,
+		     const struct node_hfsc_sc *);
+int ifdisc_lookup(struct npf_altq *);
+
+int npfdev;
+int altqsupport;
+
+TAILQ_HEAD(altqs, npf_altq) altqs = TAILQ_HEAD_INITIALIZER(altqs);
+#define is_sc_null(sc)	(((sc) == NULL) || ((sc)->m1 == 0 && (sc)->m2 == 0))
+
+struct node_queue *queues = NULL;
+
+#define FREE_LIST(T,r) \
+	do { \
+		T *p, *node = r; \
+		while (node != NULL) { \
+			p = node; \
+			node = node->next; \
+			free(p); \
+		} \
+	} while (0)
+
+#define LOOP_THROUGH(T,n,r,C) \
+	do { \
+		T *n; \
+		if (r == NULL) { \
+			r = calloc(1, sizeof(T)); \
+			if (r == NULL) \
+				err(1, "LOOP: calloc"); \
+			r->next = NULL; \
+		} \
+		n = r; \
+		while (n != NULL) { \
+			do { \
+				C; \
+			} while (0); \
+			n = n->next; \
+		} \
+	} while (0)
+int
+npfctl_test_altqsupport(int dev)
+{
+	struct npfioc_altq pa;
+	if (ioctl(dev, IOC_NPF_GET_ALTQS, &pa)) {
+		if (errno == ENODEV) {
+			fprintf(stderr, "No ALTQ support in kernel\n"
+				"ALTQ related functions disabled\n");
+			return (0);
+		} else
+		err(1, "IOC_GET_ALTQS");
+	}
+	return (1);
+}
+/* evaluate bandwidth */
+int
+npfctl_eval_bw(struct node_queue_bw * bw, char * bw_spec)
+{
+	double	 bps;
+	char	*cp;
+	bw->bw_percent = 0;
+	bps = strtod(bw_spec, &cp);
+	if (cp != NULL) {
+		if (!strcmp(cp, "b"))
+			; /* nothing */
+		else if (!strcmp(cp, "Kb"))
+			bps *= 1000;
+		else if (!strcmp(cp, "Mb"))
+			bps *= 1000 * 1000;
+		else if (!strcmp(cp, "Gb"))
+			bps *= 1000 * 1000 * 1000;
+		else if (!strcmp(cp, "%")) {
+			if (bps < 0 || bps > 100) {
+				yyerror("bandwidth spec "
+					"out of range");
+				return (-1);
+			}
+			bw->bw_percent = bps;
+			bps = 0;
+		} else {
+			yyerror("unknown unit %s", cp);
+			return (-1);
+		}
+	}
+	bw->bw_absolute = (u_int32_t)bps;
```

Comment:
if you enter just 10 for bandwidth, you hit a syntax error because of the specified rule for bandwidth (must contain a figure and a string(being byte wise or percentage). a NULL cp(string) won't even get evaluated.
 
 checking for a NULL case in the evaluation looks like a code which will be dead.